core/shared/src/main/scala/zio/AssertionFailure.scala

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+/**
+ * Represents assertion failure.
+ *
+ * This class is used to avoid constructing stack trace and special handling by `zio-test`.
+ */
+final case class AssertionFailure(message: String)
```

Comment:
It might be good to make this extend `AssertionError`.

Reply:
In fact, maybe we could just use `AssertionError`? Then one less file, one less class we have to include.

Reply:
I like that idea. I think @ioleo was looking to not have the stack trace printed to make his mocking output clean, but I think that makes it a lot less general because in a lot of cases the user wants to see the stack trace. Maybe we can just provide a helper method to remove the stack trace or change the way it is rendered. Feels like this is coming up a couple of times in this and the timeout reporting PR.

Reply:
The reasoning for this as a seperate class was to:
* not pay the cost of stack trace construction
* have special handling in `zio-test` such that assertions failures made this way are simply reported in one line (instead of cluttering output with full stack and execution trace --> which are useful in scenarios where an "unexpected" exception was throws --> in case of test assertions this is an "expected" failure --> if the predicate does not hold, we expect the test to fail)

Reply:
See also @ghostdogpr comment on [#1602](https://github.com/zio/zio/pull/1602/files#r322037212)

Reply:
@jdegoes 

1) I was only wondering, if perhaps moving `AssertionFailure` to `zio-test` and also `assert` & `assert_` to implicit class in `zio-test`, to not clutter the "main ZIO" API.

For all other use cases (that don't need special handling in `zio-test` `filterOrDie` & `filterOrDieMessage` should do the job.

2) Or maybe have `assert` & `assert_` die with `java.lang.AssertionError` as you proposed, and have implicit class in `zio-test` with `testAssert` & `testAssert_` (or whatever diffrent name) handle the "special case".

WDYT?